qualcommax: ipq60xx: add Linksys MR7350 support

Vladdrako opened the pull request:
Linksys MR7350 is a 802.11ax Dual-band router/AP.
Specifications:
* CPU: Qualcomm IPQ6000 Quad core Cortex-A53(A73) 1.5GHz
* RAM: 512MB of DDR3
* Storage: 256Mb NAND
* Ethernet: 5x1G RJ45 ports (QCA8075)
* WLAN:
* 2.4GHz: Qualcomm QCN5022 2x2 802.11b/g/n/ax 574 Mbps PHY rate
* 5GHz: Qualcomm QCN5052 2x2@80MHz 802.11a/b/g/n/ac/ax 1201 Mbps PHY rate
* LED-s:
* RGB system led
* USB blue led

* Buttons: 1x Soft reset 1x WPS
* Power: 12V DC Jack

Installation instructions:
Open Linksys Web UI - http://192.168.1.1/ca or http://10.65.1.1/ca depending on your setup.
Login with your admin password. The default password can be found on a sticker under the device.
To enter into the support mode, click on the “CA” link and the bottom of the page.
Open the “Connectivity” menu and upload the squash-factory image with the “Choose file” button.
Click start. Ignore all the prompts and warnings by click “yes” in all the popups.
The Wifi radios are turned off by default. To configure the router, you will need to connect your computer to the LAN port of the device.
Then you would need to write openwrt to the other partition for it to work
- First Check booted partition
fw_printenv -n boot_part

Change the partition first:
fw_setenv boot_part 1
or
fw_setenv boot_part 2
depending on the current partition.

- Then install Openwrt to the other partition if booted in slot 1:
mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq60xx-linksys_mr7350-squashfs-factory.bin alt_kernel

- If in slot 2:
mtd -r -e kernel -n write openwrt-qualcommax-ipq60xx-linksys_mr7350-squashfs-factory.bin kernel

Depends on https://github.com/openwrt/firmware_qca-wireless/pull/34 and https://github.com/openwrt/openwrt/pull/14815

testuser7 commented:
You should add to the instructions that before running mtd with the `-r` option you should change the partition first:
`fw_setenv boot_part 1`
or
`fw_setenv boot_part 2`
depending on the current partition.

testuser7 commented on target/linux/qualcommax/image/ipq60xx.mk:
```diff
@@ -23,3 +23,19 @@ define Device/netgear_wax214
+	IMAGE_SIZE := 75776k
+	NAND_SIZE := 256m
+	SOC := ipq6018
+	IMAGES := factory.bin
```
`IMAGES` should be defined using `+=`

Vladdrako replied:
Done

testuser7 commented on target/linux/qualcommax/image/ipq60xx.mk:
```diff
@@ -23,3 +23,19 @@ define Device/netgear_wax214
+	SOC := ipq6018
+	IMAGES := factory.bin
+	IMAGE/factory.bin  := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-ubi | linksys-image type=MR7350
+	DEVICE_PACKAGES := kmod-leds-pca963x ipq-wifi-linksys_mr7350 kmod-usb-dwc3-qcom
```
`kmod-usb-dwc3-qcom` is included by default: https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/Makefile#L15

Vladdrako replied:
Done

Vladdrako committed:
qualcommax: ipq60xx: add Linksys MR7350 support

Linksys MR7350 is a 802.11ax Dual-band router/AP.
Specifications:
* CPU: Qualcomm IPQ6000 Quad core Cortex-A53(A73) 1.5GHz
* RAM: 512MB of DDR3
* Storage: 256Mb NAND
* Ethernet: 5x1G RJ45 ports (QCA8075)
* WLAN:
	* 2.4GHz: Qualcomm QCN5022 2x2 802.11b/g/n/ax 574 Mbps PHY rate
	* 5GHz: Qualcomm QCN5052 2x2@80MHz or  802.11a/b/g/n/ac/ax 1201 Mbps PHY rate
* LED-s:
	* RGB system led
	* USB blue led

* Buttons: 1x Soft reset 1x WPS
* Power: 12V DC Jack

Installation instructions:
Open Linksys Web UI - http://192.168.1.1/ca or http://10.65.1.1/ca depending on your setup.
Login with your admin password. The default password can be found on a sticker under the device.
To enter into the support mode, click on the “CA” link and the bottom of the page.
Open the “Connectivity” menu and upload the squash-factory image with the “Choose file” button.
Click start. Ignore all the prompts and warnings by click “yes” in all the popups.
The Wifi radios are turned off by default. To configure the router, you will need to connect your computer to the LAN port of the device.
Then you would need to write openwrt to the other partition for it to work
- First Check booted partition:
fw_printenv -n boot_part

- Change the partition:
fw_setenv boot_part 1
or
fw_setenv boot_part 2
depending on the current partition

- Then install Openwrt to the other partition if booted in slot 1:
mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq60xx-linksys_mr7350-squashfs-factory.bin alt_kernel

- If in slot 2:
mtd -r -e kernel -n write openwrt-qualcommax-ipq60xx-linksys_mr7350-squashfs-factory.bin kernel

Co-Authored-by: Chukun Pan <amadeus@jmu.edu.cn>
Co-Authored-by: Robert Marko <robimarko@gmail.com>
Signed-off-by: Vladyslav Andreichykov <vladdrako007@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14807
Signed-off-by: Robert Marko <robimarko@gmail.com>

Vladdrako commented:
Update

- Fixed MAC addresses

aiamadeus commented on target/linux/qualcommax/image/ipq60xx.mk:
```diff
@@ -23,3 +23,19 @@ define Device/netgear_wax214
+	KERNEL_SIZE := 8192k
+	IMAGE_SIZE := 75776k
+	NAND_SIZE := 256m
+	SOC := ipq6018
```
Actually it's ipq6000

aiamadeus commented on target/linux/qualcommax/image/ipq60xx.mk:
```diff
@@ -23,3 +23,19 @@ define Device/netgear_wax214
 endef
 TARGET_DEVICES += netgear_wax214
+
+define Device/linksys_mr7350
```
Keep alphabetical order

Vladdrako replied:
Done

aiamadeus commented on target/linux/qualcommax/ipq60xx/base-files/etc/board.d/01_leds:
```diff
@@ -0,0 +1,20 @@
+	ucidef_set_led_netdev "lan2-port-link" "LAN2-PORT-LINK" "90000.mdio-1:01:green:lan" "lan2" "tx rx link_10 link_100 link_1000"
+	ucidef_set_led_netdev "lan3-port-link" "LAN3-PORT-LINK" "90000.mdio-1:02:green:lan" "lan3" "tx rx link_10 link_100 link_1000"
+	ucidef_set_led_netdev "lan4-port-link" "LAN4-PORT-LINK" "90000.mdio-1:03:green:lan" "lan3" "tx rx link_10 link_100 link_1000"
+	ucidef_set_led_netdev "wan-port-link" "WAN-PORT-LINK" "90000.mdio-1:04:green:wan" "wan" "tx rx link_10 link_100 link_1000"
```
This does not match the stock firmware.
Also you need to configure yellow led.

Vladdrako replied:
Done if I'm not misreading the code...

aiamadeus replied:
If I'm not mistaken, it should be like this:
```
ucidef_set_led_netdev "wan-port-activity" "WAN-PORT-ACTIVITY" "...amber:wan" "wan" "tx rx"
ucidef_set_led_netdev "wan-port-link" "WAN-PORT-LINK" "...green:wan" "wan" "link_10 link_100 link_1000"
```

aiamadeus commented on target/linux/qualcommax/ipq60xx/base-files/etc/board.d/02_network:
```diff
@@ -23,9 +27,29 @@ ipq60xx_setup_interfaces()
+	case $board in
+	linksys,mr7350)
+		wan_mac=$(macaddr_add "$(mtd_get_mac_ascii devinfo hw_mac_addr)")
+		lan_mac=$(macaddr_add "$(mtd_get_mac_ascii devinfo hw_mac_addr)" 1)
```
Are you sure this address is correct?

Vladdrako replied:
Yeah, `mtd_get_mac_ascii devinfo hw_mac_addr` provides the same address as on the router label.

aiamadeus replied:
I remember the lan/wan mac addresses are the same.
```
	linksys,mr7350)
		label_mac=$(mtd_get_mac_ascii devinfo hw_mac_addr)
		lan_mac=$label_mac
		wan_mac=$label_mac
		;;
```

Vladdrako replied:
Yeah, I was unsure about this. Thanks.

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6018-mr7350.dts:
```diff
@@ -0,0 +1,381 @@
+		ethernet1 = "/soc/dp2";
+		ethernet2 = "/soc/dp3";
+		ethernet3 = "/soc/dp4";
+		ethernet4 = "/soc/dp5";
```
No need, the mac address is not in here.

Vladdrako replied:
Done

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6018-mr7350.dts:
```diff
@@ -0,0 +1,381 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0x62>;
+        nxp,hw-blink;
```
Indentation issue

Vladdrako replied:
Done

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6018-mr7350.dts:
```diff
@@ -0,0 +1,381 @@
+		};
+
+		led_system_green: green@1 {
+            reg = <1>;
```
Indentation issue

Vladdrako replied:
Done

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6018-mr7350.dts:
```diff
@@ -0,0 +1,381 @@
+			};
+		};
+
+		qca8075_pqsgmii: ethernet-phy@5 {
```
Are you sure we need this?

Vladdrako replied:
Removed, just a remnant.

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6018-mr7350.dts:
```diff
@@ -0,0 +1,381 @@
+	status = "okay";
+
+	qcom,ath11k-calibration-variant = "Linksys-MR7350";
+	qcom,ath11k-fw-memory-mode = <2>;
```
This is incorrect, for a 512M memory device it should be 1.

Vladdrako replied:
Done

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6018-mr7350.dts:
```diff
@@ -0,0 +1,381 @@
+		};
+	};
+
+	usb-power-hog {
```
I prefer to use regulator:
```
	reg_usb_vbus: regulator-usb-vbus {
		compatible = "regulator-fixed";
		regulator-name = "usb_vbus";
		regulator-min-microvolt = <5000000>;
		regulator-max-microvolt = <5000000>;
		gpio = <&tlmm 61 GPIO_ACTIVE_LOW>;
		regulator-always-on;
	};
````

Vladdrako replied:
Done

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6018-mr7350.dts:
```diff
@@ -0,0 +1,381 @@
+};
+
+&usb3 {
+	status = "okay";
```
```
&usb3 {
	vbus-supply = <&reg_usb_vbus>;
	status = "okay";
};
```

Vladdrako replied:
Done

aiamadeus commented on target/linux/qualcommax/ipq60xx/base-files/etc/hotplug.d/firmware/11_fix_wifi_mac:
```diff
@@ -0,0 +1,18 @@
+
+case "$board" in
+	linksys,mr7350)
+		label_mac=$(mtd_get_mac_ascii devinfo hw_mac_addr)
```
This does not match the stock firmware.
```
	linksys,mr7350)
		addr=$(mtd_get_mac_ascii devinfo hw_mac_addr)
		[ "$PHYNBR" = "0" ] && macaddr_add $addr 2 > /sys${DEVPATH}/macaddress
		[ "$PHYNBR" = "1" ] && macaddr_add $addr 1 > /sys${DEVPATH}/macaddress
		;;
```

Vladdrako replied:
Done

kirdesde commented:
> 1. Need to fix MAC addresses
> 
>     2. CPUFreq is broken by APSS driver
> 
> 
> ```
> [    3.416345] cpr4_ipq807x_apss_read_fuse_data: apc_corner: speed bin = 0
> [    3.423108] cpr4_ipq807x_apss_read_fuse_data: apc_corner: CPR fusing revision = 2
> [    3.429939] cpr4_ipq807x_apss_read_fuse_data: apc_corner: CPR misc fuse value = 0
> [    3.437370] cpr4_ipq807x_apss_read_fuse_data: apc_corner: Voltage boost fuse config = 0 boost = disable
> [    3.444885] cpr3_mem_acc_init: apc: not using memory accelerator regulator
> [    3.454875] cpr4_ipq807x_apss_calculate_open_loop_voltages: apc_corner: fused      SVS: open-loop= 675000 uV
> [    3.460938] cpr4_ipq807x_apss_calculate_open_loop_voltages: apc_corner: fused      NOM: open-loop= 787500 uV
> [    3.470918] cpr4_ipq807x_apss_calculate_open_loop_voltages: apc_corner: fused    TURBO: open-loop= 850000 uV
> [    3.480730] cpr4_ipq807x_apss_calculate_open_loop_voltages: apc_corner: fused   STURBO: open-loop= 900000 uV
> [    3.490568] cpr4_ipq807x_apss_calculate_target_quotients: apc_corner: fused      SVS: quot[ 7]= 597, quot_offset[ 7]=   0
> [    3.500351] cpr4_ipq807x_apss_calculate_target_quotients: apc_corner: fused      NOM: quot[ 7]= 800, quot_offset[ 7]= 200
> [    3.511199] cpr4_ipq807x_apss_calculate_target_quotients: apc_corner: fused    TURBO: quot[ 7]= 905, quot_offset[ 7]= 105
> [    3.522139] cpr4_ipq807x_apss_calculate_target_quotients: apc_corner: fused   STURBO: quot[ 7]= 980, quot_offset[ 7]=  75
> [    3.533393] cpr3_regulator_init_ctrl: apc: Default CPR mode = closed-loop
> [    3.535816] core: _opp_supported_by_regulators: OPP minuV: 1 maxuV: 1, not supported by regulator
> [    3.550798] cpu cpu0: _opp_add: OPP not supported by regulators (864000000)
> [    3.559774] core: _opp_supported_by_regulators: OPP minuV: 2 maxuV: 2, not supported by regulator
> [    3.566406] cpu cpu0: _opp_add: OPP not supported by regulators (1056000000)
> [    3.575681] core: _opp_supported_by_regulators: OPP minuV: 3 maxuV: 3, not supported by regulator
> [    3.582552] cpu cpu0: _opp_add: OPP not supported by regulators (1320000000)
> [    3.591448] core: _opp_supported_by_regulators: OPP minuV: 4 maxuV: 4, not supported by regulator
> [    3.598436] cpu cpu0: _opp_add: OPP not supported by regulators (1440000000)
> [    3.607318] core: _opp_supported_by_regulators: OPP minuV: 5 maxuV: 5, not supported by regulator
> [    3.614321] cpu cpu0: _opp_add: OPP not supported by regulators (1512000000)
> [    3.623113] cpu cpu0: OPP table can't be empty
> ```

CPUFreq is working fine on other ipq60xx boards (wax214 for example)

The device might be missing the cpu regulator (PMIC, MP5496) and is using a cpu fixed voltage then.

Vladdrako commented:
> The device might be missing the cpu regulator (PMIC, MP5496) and is using a cpu fixed voltage then.

Yes, this device has an MP5496, but I don't know how to fix CPUFreq. Maybe @aiamadeus can help.

aiamadeus commented:
> Yes, this device has an MP5496, but I don't know how to fix CPUFreq.

The cpufreq patches are still under upstream review: https://github.com/aiamadeus/openwrt/commit/b5c0803f08ada363fee2b59d8447af751c07f68b

Vladdrako commented:
> Well, I have MR7350 and its using IPQ6018 for sure, its just a version that is fused for 1.5GHz and CPUFreq NVMEM takes care of that.

I don't know quallcomm labels their CPUs, but from chinese mr7350 review:
<details>

![image](https://github.com/openwrt/openwrt/assets/4357971/54b86299-7cea-41cd-897e-1859945762d2)

</details>

robimarko commented:
SMEM ID is quite clear so mine and I know of a couple more are IPQ6018, no idea why they have different SoC versions inside of the same model, but as long as they keep using MP5496 then everything will just work fine because CPUFreq NVMEM will populate the supported OPP-s

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6000-mr7350.dts:
```diff
@@ -0,0 +1,370 @@
+		};
+	};
+
+	reg_usb_vbus: regulator-usb-vbus {
```
You misplaced it, it should be under the gpio-keys node...

Vladdrako replied:
Oh, right, sorry. Also, should I rename dts and SOC to ipq6018 back? robimarco said it's clearly ipq6018

aiamadeus replied:
QCA replied this is ipq6000, I think it may be an older version

aiamadeus commented:
> SMEM ID is quite clear so mine and I know of a couple more are IPQ6018

I also have MR7350, the chip screen printing is ipq6000.
But oem boot log says: `CPU: IPQ6018, SoC Version: 1.0`
I tend to be this is ipq6000 different variants.

robimarko commented:
> > SMEM ID is quite clear so mine and I know of a couple more are IPQ6018
> 
> I also have MR7350, the chip screen printing is ipq6000. But oem boot log says: `CPU: IPQ6018, SoC Version: 1.0` I tend to be this is ipq6000 different variants.

Kernel is identifying it by SMEM ID, IPQ6000 plain model for sure doesnt run on 1.5GHz

aiamadeus commented:
The ipq6000 SoCs (without pmic) can force overclocking to 1.5GHz, but this makes it very susceptible to heating up. I have another ipq60xx router, the chip screen and the manufacturer's parameters all say it is ipq6000, but smem reports it is ipq6018. This can be very confusing, so I prefer to follow the hardware markup.

robimarko commented:
Yeah, silicone marking should be correct.

Vladdrako commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6000-mr7350.dts:
```diff
@@ -0,0 +1,370 @@
+
+			qcom,package-mode = "psgmii";
+
+			leds {
```
@Ansuel Help me a bit, please. Looks like I was wrong somewhere. With such LEDs config (without any scripts) green LEDs are used for 'tx tx' and yellow for 100mbit 'tx rx' (not working on 1Gbit). I tried to configure them through LuCi but the 'tx rx' mode is not working. Only 'link' and 'disabled' works.

aiamadeus replied:
Have you tried the LED 2? BTW I think it's closer to amber than yellow.

Vladdrako replied:
> Have you tried the LED 2?

What do you mean? `reg = <2>;`? It's not working. Sorry for noob questions :)

> I think it's closer to amber than yellow.

Hmm, I just copy-pasted colors from somewhere, I will change to amber later.

Vladdrako replied:
Amber LED lan1: 'tx rx' mode works if selected lan2/lan3/lan4/wan
Same for all LEDs (both green and amber). They are not working on their ports but work for others.
But the 'link' mode works correctly for both LEDs on all ports.
So, only in 'tx rx' mode ports LEDs are swapped for some reason.

aiamadeus commented:
Hi, I managed to get everything working: https://github.com/aiamadeus/openwrt/commit/8266c13f5813037b1454b957c104c023edc8a75b
Now the network port LEDs lighting method and mac address are consistent with the stock firmware.
If you don't mind, you can `git cherry-pick` my commit and add your own SoB information.
```
Signed-off-by: ...
[add wifi firmware and commit message]
Signed-off-by: ...
```

aiamadeus commented:
> I added your changes and made you as Co-Authored if you are ok with that :)

Can you just pick and modify things like include and add wifi fw?
I also fixed other style issues and I'm too lazy to comment on it.

testuser7 commented on target/linux/qualcommax/image/ipq60xx.mk:
```diff
@@ -11,6 +11,25 @@ define Device/8devices_mango-dvk
+	KERNEL_SIZE := 8192k
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	UBINIZE_OPTS := -E 5
```
Do you need to use this option?

Vladdrako replied:
I just copy-pasted it from @aiamadeus commit. Removed.

Vladdrako commented:
> Hi sorry for hijacking this thread, but is MR7350 /proc/cpuinfo showing 0x51 implementer and 0x801 part?

Yes
```
root@MR7350:~# cat /proc/cpuinfo
processor       : 0
BogoMIPS        : 48.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
CPU implementer : 0x51
CPU architecture: 8
CPU variant     : 0xa
CPU part        : 0x801
CPU revision    : 4
```

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6000-mr7350.dts:
```diff
@@ -0,0 +1,375 @@
+		lan1-amber {
+			label = "amber:lan1";
+			color = <LED_COLOR_ID_AMBER>;
+			gpios = <&qca8075_0 0 GPIO_ACTIVE_HIGH>;
```
Why not drive these as netdev LED-s as that is offloaded?

Vladdrako replied:
Sorry, but can you give me an example? I don't know how to do it. My knowledge is limited.

testuser7 replied:
Take look here: https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8071-ax3600.dts#L76

Vladdrako replied:
Yeah, I tried this before. https://github.com/openwrt/openwrt/pull/14807#discussion_r1518886785

robimarko replied:
I will give it a go since there is no reason why they should not work unless Linksys wired them incorrectly

robimarko commented on target/linux/qualcommax/ipq60xx/base-files/etc/board.d/02_network:
```diff
@@ -26,9 +29,30 @@ ipq60xx_setup_interfaces()
+
+	case $board in
+	linksys,mr7350)
+		label_mac=$(mtd_get_mac_ascii devinfo hw_mac_addr)
```
Is this a u-boot env like setup?

testuser7 replied:
The `devinfo` partition contains key=value data:
```
hexdump -C -n 48 /dev/mtd20
00000000  6d 66 67 5f 64 61 74 61  5f 76 65 72 73 69 6f 6e  |mfg_data_version|
00000010  3d 31 0a 6d 6f 64 65 6c  4e 75 6d 62 65 72 3d 4d  |=1.modelNumber=M|
00000020  58 38 35 30 30 0a 73 65  72 69 61 6c 5f 6e 75 6d  |X8500.serial_num|
00000030
```

robimarko replied:
Yeah, sadly no luck with NVMEM currently then

robimarko commented:
BTW, is the WLAN.HK.2.5.0.1-03982-QCAHKSWPL_SILICONZ-3 FW working on this device?

Cause I get:
```
[   13.101356] ath11k c000000.wifi: ipq6018 hw1.0
[   13.101390] ath11k c000000.wifi: FW memory mode: 1
[   13.104961] remoteproc remoteproc0: powering up cd00000.remoteproc
[   13.109541] remoteproc remoteproc0: Booting fw image IPQ6018/q6_fw.mdt, size 7340
[   13.170356] urngd: v1.0.2 started.
[   13.408757] remoteproc remoteproc0: remote processor cd00000.remoteproc is now up
[   13.410064] ath11k c000000.wifi: qmi ignore invalid mem req type 3
[   13.415684] ath11k c000000.wifi: chip_id 0x0 chip_family 0x4 board_id 0xff soc_id 0xffffffff
[   13.421327] ath11k c000000.wifi: fw_version 0x25008f8e fw_build_timestamp 2024-03-01 03:54 fw_build_id WLAN.HK.2.5.0.1-03982-QCAHKSWPL_SILICONZ-3
[   23.529764] ath11k c000000.wifi: failed to wait wlan mode request (mode 7): -110
[   23.529815] ath11k c000000.wifi: qmi failed to send wlan fw mode: -110
[   29.077408] qcom-q6v5-wcss-pil cd00000.remoteproc: watchdog received: 
[   29.077408] QC Image Version: QC_IMAGE_VERSION_STRING=WLAN.HK.2.5.0.1-03982-QCAHKSWPL_SILICONZ-3
[   29.077408] Image Variant : IMAGE_VARIANT_STRING=6018.wlanfw.evalQ
[   29.077408] 
[   29.077408]     :Excep  :0 Exception detectedparam0 :zero, param1 :zero, param2 :zero.
[   29.077408] Thread ID      : 0x00000061  Thread name    : WLAN_SCHED1  Process ID     : 0
[   29.077408] Register:
[   29.077408] SP : 0x4bb89a28
[   29.077408] FP : 0x4bb89a40
[   29.077408] PC : 0x4ad9c074
[   29.077408] SSR : 0x00000003
[   29.077408] BADVA : 0x00000004
[   29.077408] LR : 0x4ad9bf5c
[   29.077408] 
[   29.077408] Stack Dump
[   29.077408] from : 0x4bb89a28
[   29.077408] to   : 0x4bb89bc0
[   29.077408] 
[   29.123613] remoteproc remoteproc0: crash detected in cd00000.remoteproc: type watchdog
[   29.145839] remoteproc remoteproc0: handling crash #1 in cd00000.remoteproc
[   29.153803] remoteproc remoteproc0: recovering cd00000.remoteproc
[   29.201645] remoteproc remoteproc0: stopped remote processor cd00000.remoteproc
[   29.494869] remoteproc remoteproc0: remote processor cd00000.remoteproc is now up
[   29.496204] ath11k c000000.wifi: qmi ignore invalid mem req type 3
[   29.501888] ath11k c000000.wifi: chip_id 0x0 chip_family 0x4 board_id 0xff soc_id 0xffffffff
[   29.507440] ath11k c000000.wifi: fw_version 0x25008f8e fw_build_timestamp 2024-03-01 03:54 fw_build_id WLAN.HK.2.5.0.1-03982-QCAHKSWPL_SILICONZ-3
[   39.529757] ath11k c000000.wifi: failed to wait wlan mode request (mode 7): -110
[   39.529805] ath11k c000000.wifi: qmi failed to send wlan fw mode: -110
```

mrnuke commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6000-mr7350.dts:
```diff
@@ -0,0 +1,375 @@
+	reset-gpios = <&tlmm 75 GPIO_ACTIVE_LOW>;
+
+	ethernet-phy-package@0 {
+		compatible = "qcom,qca8075-package";
```
`qcom,package-mode` ?

testuser7 replied:
`psgmi` is the default mode so I guess there is no need to specify.

mrnuke replied:
I understand it's the default. I should have explained my reasoning in the first comment. It's not obvious to a casual observer that there is a package mode, and that it has a value of "psgmii". I don't think anyone who is not familiar with qca8075 would expect this.

So while "psgmii" is the default, I think being explicit that there is a "package-mode", and it is set to "psgmii" helps readability. @Ansuel or @robimarko  might choose to override me on this.

robimarko replied:
Either is correct as bindings dont require package-mode property to be set and the default is documented:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/tree/Documentation/devicetree/bindings/net/qcom,qca807x.yaml?h=next-20240419

That being said, I dont have anything against it being explicitly set.

Vladdrako commented on target/linux/qualcommax/ipq60xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -33,6 +33,20 @@ EOF
 platform_do_upgrade() {
 	case "$(board_name)" in
+	linksys,mr7350)
+		boot_part="$(fw_printenv -n boot_part)"
```
What about parsing cmdline like that
```
		if [[ $(cat /proc/cmdline) =~ 'ubi.mtd=rootfs' ]]; then
			fw_setenv boot_part 2
			CI_KERNPART="alt_kernel"
			CI_UBIPART="alt_rootfs"
		else
			fw_setenv boot_part 1
			CI_UBIPART="rootfs"
		fi
		fw_setenv boot_part_ready 3
		fw_setenv auto_recovery yes
		nand_do_upgrade "$1"
```

testuser7 replied:
I think that sometimes we may intentionally want to change this variable in order to update the current partition.
I'm not sure if a change is needed here.

Vladdrako replied:
@testuser7 Yeah, agree. And I just copy-pasted the code from ipq807x
https://github.com/openwrt/openwrt/blob/e4453d1f8172feee97e0f66a102aa4e63c47e629/target/linux/qualcommax/ipq807x/base-files/lib/upgrade/platform.sh#L81

mrnuke replied:
@testuser7, I think you  misunderstand my comment. I suggested to not read `boot_part`. It obviously need to be written to ping-pong the boot partition.

@Vladdrako, Thank you for pointing that out. I think that copypasting code is a big no-no because it duplicates any mistakes and issues that already exist, and it creates redundant code.

The preferred procedure would be to share the common functionality. One way is to move this logic to `qualcommax/base-files/` in something named `linksys.sh` and source this file as needed.

if you integrate all my suggestions, you get to something that is symmetrical and doesn't wear out the NAND:

		if <your heuristic here>; then
			new_boot_part=2
			CI_KERNPART="alt_kernel"
			CI_UBIPART="alt_rootfs"
		else
			new_boot_part=1
			CI_KERNPART="kernel"
			CI_UBIPART="rootfs"
		fi
		
		fw_setenv -s - <<-EOF
			boot_part $new_boot_part
			boot_part_ready 3
			auto_recovery yes
		EOF

		nand_do_upgrade "$1"

testuser7 replied:
> @testuser7, I think you misunderstand my comment. I suggested to not read `boot_part`. It obviously need to be written to ping-pong the boot partition.

I don't think there's anything wrong with reading the u-boot variable. If someone wants to, they can overwrite them so much that the system won't even start.

mrnuke replied:
> I don't think there's anything wrong with reading the u-boot variable. If someone wants to, they can overwrite them so much that the system won't even start.

This has nothing to do with my original comment about using `boot_part` as the heuristic for the installation destination.

testuser7 commented on target/linux/qualcommax/ipq60xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -33,6 +33,20 @@ EOF
+			CI_UBIPART="alt_rootfs"
+		else
+			fw_setenv boot_part 1
+			CI_UBIPART="rootfs"
```
`kernel` is a default value for `CI_KERNPART` variable.

mrnuke replied:
I understand it's the default. My issue is with the cognitive dissonance by omitting it. Someone looking at the code can reasonably ask:

- Why is`CI_KERNPART` omitted for the first partition?
- Is the kernel only stored in the second partition?

Unless someone is intimately familiar with the implementation details of `nand_do_upgrade` this piece of code here is horrendously confusing. By being explicit about the values of `CI_{KERN,UBI}PART` in both cases, it becomes obvious that there is a symmetry, that both partitions are relevant and important.

This is about redability, and avoiding micro-optimizations that hurt it.

mrnuke commented on target/linux/qualcommax/ipq60xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -33,6 +33,20 @@ EOF
+			CI_UBIPART="rootfs"
+		fi
+		fw_setenv boot_part_ready 3
+		fw_setenv auto_recovery yes
```
This is overkill.

```
		if [ "$(fw_printenv -n boot_part_ready)" -ne "3" ] ; then
			echo "boot_part_ready 3" >> $setenv_script
		fi
```
The entire NAND block will need to be rewritten. Whether an individual variable is modified or not does not matter. So this form adds a lot of code with no real benefit.

mrnuke replied:
I do not have this issue in my sysupgrade branch [1]. Is the OEM UBI volume removed so that there is enough space for `rootfs_data?

[1] https://github.com/mrnuke/openwrt/blob/2d2febc896bd61ed3371db70b53566f06ab5bc38/target/linux/qualcommax/ipq60xx/base-files/lib/upgrade/platform.sh#L130

testuser7 replied:
Do you have single `rootfs_data` partition?
For other Linksys devices we have two separate `rootfs_data` partitions: https://forum.openwrt.org/t/adding-openwrt-support-for-linksys-mx5300/92667/25

mrnuke replied:
Nope. One `rootfs_data` per UBI volume.

What is the output of  `ubinfo -a /dev/ubi0` ?

Vladdrako replied:
> What is the output of `ubinfo -a /dev/ubi0` ?

```
ubi0
Volumes count:                           2
Logical eraseblock size:                 126976 bytes, 124.0 KiB
Total amount of logical eraseblocks:     592 (75169792 bytes, 71.6 MiB)
Amount of available logical eraseblocks: 0 (0 bytes)
Maximum count of volumes                 128
Count of bad physical eraseblocks:       0
Count of reserved physical eraseblocks:  40
Current maximum erase counter value:     3
Minimum input/output unit size:          2048 bytes
Character device major/minor:            245:0
Present volumes:                         0, 1

Volume ID:   0 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        326 LEBs (41394176 bytes, 39.4 MiB)
State:       OK
Name:        rootfs
Character device major/minor: 245:1
-----------------------------------
Volume ID:   1 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        222 LEBs (28188672 bytes, 26.8 MiB)
State:       OK
Name:        rootfs_data
Character device major/minor: 245:2
```

mrnuke replied:
That looks as expected. Thank you!

testuser7 replied:
> Nope. One `rootfs_data` per UBI volume.

But you have two `rootfs` partitions:
- `rootfs` - mtd15
- `alt_rootfs` - mtd17
And after sysupgrade you are switched between mtd15 and mtd17.
Each `rootfs` partition contains only one `rootfs_data` volume.

testuser7 replied:
@Vladdrako can you show dmesg output how the partitions are created?
`Creating 20 MTD partitions on "qcom_nand.0":`

mrnuke replied:
Yes, that is exactly what I would expect.

Why do you believe that having separate `rootfs_data` on separate UBI devices is a problem, and why do you believe this would prevent config from being saved on sysupgrade?

Vladdrako replied:
> can you show dmesg output how the partitions are created?

```
[    1.415891] 0x000000000000-0x000000180000 : "0:sbl1"
[    1.423027] 0x000000180000-0x000000280000 : "0:mibib"
[    1.427497] 0x000000280000-0x000000600000 : "0:qsee"
[    1.434178] 0x000000600000-0x000000680000 : "0:devcfg"
[    1.437105] 0x000000680000-0x000000700000 : "0:rpm"
[    1.442004] 0x000000700000-0x000000780000 : "0:cdt"
[    1.446832] 0x000000780000-0x000000800000 : "0:appsblenv"
[    1.451729] 0x000000800000-0x000000980000 : "0:appsbl"
[    1.457942] 0x000000980000-0x000000a00000 : "0:art"
[    1.462285] 0x000000a00000-0x000001300000 : "0:wififw"
[    1.473076] 0x000001300000-0x000001380000 : "0:ethphyfw"
[    1.473937] 0x000001380000-0x000001400000 : "u_env"
[    1.478274] 0x000001400000-0x000001440000 : "s_env"
[    1.482701] 0x000001440000-0x000001480000 : "devinfo"
[    1.487597] 0x000001480000-0x000001c80000 : "kernel"
[    1.498296] 0x000001c80000-0x000006680000 : "rootfs"
[    1.551163] mtd: setting mtd15 (rootfs) as root device
[    1.551458] mtdsplit: no squashfs found in "rootfs"
[    1.555244] 0x000006680000-0x000006e80000 : "alt_kernel"
[    1.566141] 0x000006e80000-0x00000b880000 : "alt_rootfs"
[    1.620480] 0x00000b880000-0x00000ba80000 : "sysdiag"
[    1.622480] 0x00000ba80000-0x00000fe80000 : "syscfg"
```

testuser7 replied:
> ```
> [    1.487597] 0x000001480000-0x000001c80000 : "kernel"
> [    1.498296] 0x000001c80000-0x000006680000 : "rootfs"
> [    1.555244] 0x000006680000-0x000006e80000 : "alt_kernel"
> [    1.566141] 0x000006e80000-0x00000b880000 : "alt_rootfs"
> ```

There are some differences here. In OEM firmware, the `kernel` partition overlaps the `rootfs` partition:
```
[    1.155578] 0x000001480000-0x000006680000 : "kernel"
[    1.218653] 0x000001c80000-0x000006680000 : "rootfs"
[    1.276153] 0x000006680000-0x00000b880000 : "alt_kernel"
[    1.339189] 0x000006e80000-0x00000b880000 : "alt_rootfs"
```
Keeping this layout will help in reverting to OEM firmware.

testuser7 replied:
> Yes, that is exactly what I would expect.
> 
> Why do you believe that having separate `rootfs_data` on separate UBI devices is a problem, and why do you believe this would prevent config from being saved on sysupgrade?

So every single file (even one created manually) is copied to another `rootfs_data` volume on the second `rootfs` partition during sysupgrade?

mrnuke replied:
My expectation is that an upgrade copies all relevant configuration to the `rootfs_data` volume of the UBI device given in `CI_UBIPART`. If that isn't happening,let me know what you observe, and I can help troubleshoot.

Note that this doesn't mean "every single file". sysupgrade shouldn't copy everything blindly.

robimarko replied:
You are supposed to have `rootfs_data` per each rootfs UBI volume, that is normal.

Vladdrako commented on target/linux/qualcommax/ipq60xx/base-files/etc/board.d/02_network:
```diff
@@ -26,9 +29,30 @@ ipq60xx_setup_interfaces()
+	local board="$1"
+	local lan_mac=""
+	local wan_mac=""
+	local label_mac=""
```
What about this
```
ipq60xx_setup_macs()
{
	local board="$1"

	case "$board" in
		linksys,mr7350)
			hw_mac_addr=$(mtd_get_mac_ascii devinfo hw_mac_addr)
			ucidef_set_interface_macaddr "lan" "$hw_mac_addr"
			ucidef_set_interface_macaddr "wan" "$hw_mac_addr"
		;;
	esac
}
```
?

testuser7 replied:
It can be seen that two approaches are practiced:
- https://github.com/openwrt/openwrt/blob/main/target/linux/ipq40xx/base-files/etc/board.d/02_network#L157
- https://github.com/openwrt/openwrt/blob/main/target/linux/ipq806x/base-files/etc/board.d/02_network#L75

mrnuke replied:
@Vladdrako, the above looks much clearer.

testuser7 commented on target/linux/qualcommax/image/ipq60xx.mk:
```diff
@@ -11,6 +11,24 @@ define Device/8devices_mango-dvk
+	IMAGE/factory.bin := append-kernel | pad-to $$$$(KERNEL_SIZE) | \
+		append-ubi | linksys-image type=MR7350
+	DEVICE_PACKAGES := ipq-wifi-linksys_mr7350 \
+		kmod-leds-pca963x kmod-usb-ledtrig-usbport
```
They are default for the whole target: https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/Makefile#L15

Vladdrako commented on target/linux/qualcommax/ipq60xx/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -0,0 +1,18 @@
+	linksys,mr7350)
+		addr=$(mtd_get_mac_ascii devinfo hw_mac_addr)
+		[ "$PHYNBR" = "0" ] && macaddr_add $addr 2 > /sys${DEVPATH}/macaddress
+		[ "$PHYNBR" = "1" ] && macaddr_add $addr 1 > /sys${DEVPATH}/macaddress
```
Yeah, I can use `macaddr_add $(mtd_get_mac_ascii devinfo hw_mac_addr) $(($PHYNBR + 1)) > /sys${DEVPATH}/macaddres` but then MACs will be swapped. We need +2 for phy0 and +1 for phy1 to match stock firmware.
The same uses in many targets
https://github.com/openwrt/openwrt/blob/66019e456f60f5ef71a9af0d73502e20b874d114/target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac#L29
https://github.com/openwrt/openwrt/blob/66019e456f60f5ef71a9af0d73502e20b874d114/target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac#L90

Vladdrako commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6000-mr7350.dts:
```diff
@@ -0,0 +1,375 @@
+&dp1 {
+	status = "okay";
+	phy-handle = <&qca8075_0>;
+	label = "lan1";
```
You mean using eth0..eth5 instead of lan/wan? But this is OpenWRT defaults and most scripts parse/use lan/wan if I'm not mistaken. This is much more convenient for the home router.

mrnuke replied:
I was taught the "label=" should match labeling on the physical device. Maybe that rule has changed since. @robimarko  ?

robimarko replied:
Yes, it should resemble the physical device usually.
But I am not sure how useful is calling LAN ports Ethernet1...Ethernet4 and WAN one Internet more useful than sticking with the common lanX-wan naming

oyvindln commented:
Any further progress on this?

robimarko commented:
@Vladdrako Any chance you can rebase this?

Vladdrako commented:
@robimarko Done.

robimarko commented:
Please fix the formal issue

Vladdrako commented:
@robimarko I should add a real name to my GitHub profile, right? Anyway, rebased.

castiel652 commented:
> @robimarko I should add a real name to my GitHub profile, right? Anyway, rebased.

Just in the commit sign off should be enough. :)

Vladdrako commented:
@robimarko I signed the commit, check, please. Sorry, I'm new to git :)

raenye commented:
@backslashxx O/T, but what is this nice colorful `/etc/banner` in the screenshot? a new feature in 24.10?

robimarko committed:
qualcommax: refresh config

Refresh the kernel config via kernel_menuconfig.

Link: https://github.com/openwrt/openwrt/pull/14807
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: enable userspace regulator for ipq60xx as well

Userspace regulator consumer is rather usefull so enable it for ipq60xx
as well and remove from ipq807x only kernel config.

Link: https://github.com/openwrt/openwrt/pull/14807
Signed-off-by: Robert Marko <robimarko@gmail.com>

aiamadeus commented:
![image](https://github.com/user-attachments/assets/a1b5fcba-fb06-4d2f-9c24-47053de697bd)
This doesn't look quite right.

sLTommy commented:
So no support yet? :(

mightymos commented:
> So no support yet? :(

You can try this build, however it is an unofficial third party build:
https://acidslide.geekdude.com/BackspaceWRT/latest-build/full-build-20240805/targets/qualcommax/ipq60xx/

sLTommy commented:
> > So no support yet? :(
> 
> You can try this build, however it is an unofficial third party build: https://acidslide.geekdude.com/BackspaceWRT/latest-build/full-build-20240805/targets/qualcommax/ipq60xx/

I'm using that one but it no longer works since they changed the packages to ipk

backslashxx commented:
I recommend building your own. Not only because of the experience, but access to a very high level of customizability.

you can cherry pick this over current head and it works.

example 
https://github.com/backslashxx/openwrt/commits/SNAPSHOT-MR7350/

robimarko commented:
I am working on this, just got the PHY LED-s offloaded properly so its close to being finished.

robimarko commented:
Hm, I am looking at the installation instructions and its bit confusing.

This thing has kernel, rootfs and then alt_kernel and alt_rootfs partitions but they all contain squashfs, what is the deal?

robimarko commented:
Thanks! Rebased on top of main and merged!

ruby10834343dfgfgf commented:
Hi all , I am using partion 2 with BackspaceOpenwrt version [,](https://cidslide.geekdude.com/BackspaceWRT/latest-build/full-build-20240805/targets/qualcommax/ipq60xx)

How to install this version 24.10-snapshots and it has enough 256mb memory ?

neheb commented:
In addition to the bogus install instructions, how is the USB supposed to be turned off?

I get

```
~# uhubctl -a off
No compatible devices detected!
Run with -h to get usage info.
```

which is hilarious that there's a regulator specified for the USB interface.

robimarko commented:
What do you mean by bogus install instructions?
Those worked just fine for me.

As for USB, it was you who switched it from the user-controllable userspace regulator to the `vdda-phy-supply` on the USB PHY
https://github.com/openwrt/openwrt/commit/1306237fab40781f01a5635f1688d28fc68b77f4

neheb commented:
```
Then install Openwrt to the other partition if booted in slot 1:
mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq60xx-linksys_mr7350-squashfs-factory.bin alt_kernel

If in slot 2:
mtd -r -e kernel -n write openwrt-qualcommax-ipq60xx-linksys_mr7350-squashfs-factory.bin kernel
```

This caused a brick. The reason being that the partition is 8MB.

Pepermint98 commented:
Hello all, first github post, hope to help out with the linksys mr7350.

Bricking confirmed and solution found.  Three people, including myself have flashed the Linksys MR7350 successfully as of this writing using the method referenced (08 December 2025).  i am unsure the best way to address this issue since I am not a coder.

- openwrt thread
     https://forum.openwrt.org/t/linksys-mr7350-issues/240379/74

- issue and solution here
     https://github.com/openwrt/openwrt/issues/19261#issuecomment-3482216199

dd if=openwrt-factory.bin bs=8M count=1 of=kernel.bin
dd if=openwrt-factory-bin bs=8M skip=1 of=rootfs.bin
mtd -e kernel -n write kernel.bin kernel
mtd -e rootfs -n write rootfs.bin rootfs
fw_setenv boot_part 1
reboot

lorand-horvath commented:
Take a look at this PR, which might have fixed the sysupgrade issues noticed on `ipq60xx` and `ipq807x`: https://github.com/openwrt/openwrt/pull/21760
